auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/internal/EditAccessGroupsController.kt

```diff
@@ -0,0 +1,241 @@
+        callingUser: LdapUser,
+        selectedOrgs: Set<String>
+    ) {
+        val userDomainOrgs =
+            if (callingUser.email == null) listOf() else organisationService.findAllByDomain(callingUser.email)
+                .map { it.code }
+        for (org in selectedOrgs) {
+            if (!userDomainOrgs.contains(org)) {
+                throw ApiError(
+                    HttpStatusCode.Forbidden,
+                    "external_user_non_online_registration_group",
+                    "User attempted to assign self to an organisation not in their domain: $org",
+                )
+            }
+        }
+    }
+
+    private fun validateAccessGroupRequest(
+        accessGroupsRequested: Map<String, List<String>>,
+        allAccessGroups: Map<String, AccessGroup>,
+        selectedOrgs: Set<String>,
+        userIsInternal: Boolean
+    ) {
+        for (requestedAccessGroup in accessGroupsRequested) {
+            val accessGroupData = allAccessGroups.getOrElse(requestedAccessGroup.key) {
+                throw ApiError(
+                    HttpStatusCode.Forbidden,
+                    "inexistent_group",
+                    "Attempted to assign a user an access group that does not exist: "
+                            + requestedAccessGroup.key,
+                )
+            }
+
+            if (userIsInternal && !accessGroupData.enableInternalUser) {
+                throw ApiError(
+                    HttpStatusCode.Forbidden,
+                    "internal_user_non_internal_group",
+                    "Attempted to assign an internal user a group not enabled for internal users: "
+                            + requestedAccessGroup.key,
+                )
+            }
+
+            if (!userIsInternal && !accessGroupData.enableOnlineRegistration) {
+                throw ApiError(
+                    HttpStatusCode.Forbidden,
+                    "external_user_non_online_registration_group",
+                    "Attempted to assign an external user a group not enabled for online registration: "
+                            + requestedAccessGroup.key,
+                )
+            }
+
+            for (requestedGroupOrg in requestedAccessGroup.value) {
+                if (!selectedOrgs.contains(requestedGroupOrg)) {
+                    throw ApiError(
+                        HttpStatusCode.Forbidden,
+                        "external_user_non_online_registration_group",
+                        "Attempted to assign user access group an organisation not in their selected organisations: $requestedGroupOrg"
+                    )
+                }
+            }
+        }
+    }
+
+    public fun generateAccessGroupActionList(
+        accessGroupsRequestMap: Map<String, List<String>>,
+        currentAccessGroups: Map<String, List<String>>,
+        selectedOrgs: Set<String>
+    ): MutableList<AccessGroupAction> {
+        val accessGroupActions = mutableListOf<AccessGroupAction>()
```

Comment:
`public` is the default, we could use set here, and we don't need to return it as mutable